tools/run_tests/xds_k8s_test_driver/tests/authz_test.py

```diff
@@ -0,0 +1,313 @@
+            {
+                "destinations": {
+                    "hosts": [
+                        f"not-the-host:{self.server_xds_port}",
+                        "not-the-host",
+                    ],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-host",
+                    },
+                },
+            },
+            {
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [1],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-port",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": ["*"],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "principal-present",
+                    },
+                },
+            },
+            {
+                "sources": [{
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                }, {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                        f"spiffe://{self.project}.svc.id.goog/ns/"
+                        f"{self.client_namespace}/sa/{self.client_name}",
+                    ],
+                }],
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "match-principal",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-principal",
+                    },
+                },
+            },
+        ]
+
+    def configure_and_assert(self, test_client: _XdsTestClient,
+                             test_metadata_val: Optional[str],
+                             status_code: grpc.StatusCode) -> None:
+        # Swap method type every sub-test to avoid mixing results
```

Comment:
Just curious: Why will this happen? Is this a potential race that previous config might leak to next testing session? We should be examining the diff, in theory?

Test client should live for the entire life span of a test case. So, all `configure_and_assert` in a test case (not sub test) is sharing one test client.

Reply:
> If we want absolute isolation, we should promote subtests to tests

How do subtests provide absolute isolation? I didn't realize there was _any_ isolation between subtests, except for exception-throwing.

Normal flow:
1. Test client is doing Unary RPCs with some arbitrary configuration
2. Reconfigure client, and swap to Empty RPCs. Since there have not been any recent empty RPCs, all empty RPCs from this point forward are using our new configuration
3. Get baseline stats. Some Unary RPCs may still be happening with the old config, but we don't care
4. Get more stats, Do diff. Only look at _recent_ empty RPCs
5. Restart at 1 for next test, but swapping the role of Empty and Unary RPCs

> Will stats.stats_per_method[t].rpcs_started become zero if we change configuration?

No. We get a zero-like baseline by subtracting the stats at the beginning of the test.

Reply:
This uses the security-enabled setup. I used a visibility-restricted discovery file, but just for the normal security stuff (which included RBAC stuff, just not for gRPC). And you can only test against staging TD at this precise moment.

But that precise error is "there was an error and our parsing code crashes trying to figure out what the error was." I just put a print() statement in that error-handling code to print things out before we try to convert the response. Then I can see the true error. See also b/191758513.